ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/BatchingLedgerWriter.scala

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.api
+
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicBoolean
+
+import akka.stream.scaladsl.{Keep, Sink, Source, SourceQueueWithComplete}
+import akka.stream.{Materializer, OverflowStrategy, QueueOfferResult}
+import com.daml.ledger.participant.state.kvutils
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.DamlSubmissionBatch
+import com.daml.ledger.participant.state.kvutils.Envelope
+import com.daml.ledger.participant.state.v1.{ParticipantId, SubmissionResult}
+import com.digitalasset.ledger.api.health.HealthStatus
+import com.digitalasset.logging.LoggingContext.newLoggingContext
+import com.digitalasset.logging.{ContextualizedLogger, LoggingContext}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+
+trait BatchingQueue {
+  def run(commitBatch: Seq[DamlSubmissionBatch.CorrelatedSubmission] => Future[Unit])(
+      implicit materializer: Materializer): BatchingQueueHandle
+}
+
+trait BatchingQueueHandle {
+  def alive: Boolean
+  def offer(submission: DamlSubmissionBatch.CorrelatedSubmission): Future[SubmissionResult]
+}
+
+/** Default batching queue implementation for the batching ledger writer.
+  *
+  * @param maxQueueSize The maximum number of submissions to queue for batching. On overflow new submissions are dropped.
+  * @param maxBatchSizeBytes Maximum size for the batch. A batch is emitted if adding a submission would exceed this limit.
+  * @param maxWaitDuration The maximum time to wait before a batch is forcefully emitted.
+  * @param maxConcurrentCommits The maximum number of concurrent calls to make to [[LedgerWriter.commit]].
+  */
+case class DefaultBatchingQueue(
```

Comment:
Please move this into a separate file along with the above trait definitions or move the traits into a separate file as well. This file should only contain the `BatchingLedgerWriter`.

Reply:
Done.